app/domain/authentication/authn_oidc/pkce_support_feature/client.rb

```diff
@@ -0,0 +1,117 @@
+module Authentication
+  module AuthnOidc
+    module PkceSupportFeature
+      class Client
+        def initialize(
+          authenticator:,
+          client: ::OpenIDConnect::Client,
+          oidc_id_token: ::OpenIDConnect::ResponseObject::IdToken,
+          discovery_configuration: ::OpenIDConnect::Discovery::Provider::Config,
+          cache: Rails.cache,
+          logger: Rails.logger
+        )
+          @authenticator = authenticator
+          @client = client
+          @oidc_id_token = oidc_id_token
+          @discovery_configuration = discovery_configuration
+          @cache = cache
+          @logger = logger
+        end
+
+        def oidc_client
+          @oidc_client ||= begin
+            issuer_uri = URI(@authenticator.provider_uri)
+            @client.new(
+              identifier: @authenticator.client_id,
+              secret: @authenticator.client_secret,
+              redirect_uri: @authenticator.redirect_uri,
+              scheme: issuer_uri.scheme,
+              host: issuer_uri.host,
+              port: issuer_uri.port,
+              authorization_endpoint: URI(discovery_information.authorization_endpoint).path,
+              token_endpoint: URI(discovery_information.token_endpoint).path,
+              userinfo_endpoint: URI(discovery_information.userinfo_endpoint).path,
+              jwks_uri: URI(discovery_information.jwks_uri).path,
+              end_session_endpoint: URI(discovery_information.end_session_endpoint).path
+            )
+          end
+        end
+
+        def callback(code:, nonce:, code_verifier:)
+          oidc_client.authorization_code = code
+          begin
+            bearer_token = oidc_client.access_token!(
+              scope: true,
+              client_auth_method: :basic,
+              code_verifier: code_verifier
+            )
+          rescue Rack::OAuth2::Client::Error => e
+            # Only handle the expected errors related to access token retrieval.
+            case e.message
+            when /PKCE verification failed/
+              raise Errors::Authentication::AuthnOidc::TokenRetrievalFailed,
+                    'PKCE verification failed'
+            when /The authorization code is invalid or has expired/
+              raise Errors::Authentication::AuthnOidc::TokenRetrievalFailed,
+                    'Authorization code is invalid or has expired'
+            when /Code not valid/
+              raise Errors::Authentication::AuthnOidc::TokenRetrievalFailed,
+                    'Authorization code is invalid'
+            end
+            raise e
+          end
+          id_token = bearer_token.id_token || bearer_token.access_token
+
+          begin
+            attempts ||= 0
+            decoded_id_token = @oidc_id_token.decode(
+              id_token,
+              discovery_information.jwks
+            )
+          rescue Exception => e
```

Comment:
Code climate also flags this. Do we need to rescue `Exception` here, or can it be just `StandardError`? If `Exception` is intentional, mind adding a comment why that is?